cpp/ParseFasm.cpp

```diff
@@ -0,0 +1,414 @@
+};
+
+// This output stream operator adds the header needed to decode
+// a string of unknown length and type.
+// Note that some characters are escaped in hex mode to
+// avoid confusion.
+std::ostream &operator<<(std::ostream &s, const Str &str) {
+  s << str.tag << Num(str.data.size());
+  if (hex_mode) { // escape < \ >
+    for (char c : str.data) {
+      if (c == '<' || c == '>' || c == '\\') {
+        s.put('\\');
+      }
+      s.put(c);
+    }
+  } else {
+    s << str.data;
+  }
+  return s;
+}
+
+// Wraps a string in another header, used to aggregate data.
+std::string withHeader(char tag, std::string data) {
+  std::ostringstream header;
+  header << tag << Num(data.size());
+  return header.str() + data;
+}
+
+// Counts characters that don't match the given character.
+// Used to count digits skipping '_'.
+int count_without(std::string::iterator start,
+                  std::string::iterator end,
+                  char c) {
+  int count = 0;
+  auto it = start;
+  while (it != end) {
+    if (*it != c) {
+      count++;
+    }
+    it++;
+  }
+  return count;
+}
+
+// Calculates the number of leading pad bits are needed
+// so that the rightmost bit will be the LSB of a Num.
+// e.g. This would be 31 for 33'b0.
+int lead_bits(int bits)
+{
+  return (Num::width - (bits % Num::width)) % Num::width;
+}
+
+// Decode a hex digit.
+int from_hex_digit(char c) {
+  int result =
+    c >= '0' && c <= '9' ? c - '0' :
+    c >= 'a' && c <= 'f' ? c - 'a' :
+    c >= 'A' && c <= 'F' ? c - 'A' : -1;
+  assert(result >= 0 && result < 16);
+  return result;
+}
+
+
+// FasmParserBaseVisitor is a visitor for the parse tree
+// generated by the ANTLR parser.
+// It will encode the tree a line at a time and stream out of
+// the given std::ostream
+class FasmParserBaseVisitor : public FasmParserVisitor {
+
+public:
+
+  static constexpr size_t kHeaderSize = 5;
+
+
+  // The constructor requires a std::ostream to stream encoded lines.
+  // This is to avoid storing an entire copy of the parse tree in a
+  // different form.
+  FasmParserBaseVisitor(std::ostream &out) : out(out) {}
+
+  // Stream out FASM lines.
+  virtual Any visitFasmFile(FasmParser::FasmFileContext *context) override {
+    for (auto &line : context->fasmLine()) {
+      std::string str = visit(line);
+      if (!str.empty()) {
+        out << str;
+        if (hex_mode) out << std::endl;
+      }
+    }
+    return {};
+  }
+
+// Helper macro to convert a rule context into a string
+#define GET(x) (context->x() ? visit(context->x()).as<std::string>() : "")
+
+  // This is called for each FASM line.
+  // Tags: comment (#), line (l)
+  virtual Any visitFasmLine(FasmParser::FasmLineContext *context) override {
+    std::ostringstream data;
+    data << GET(setFasmFeature)
+         << GET(annotations);
+
+    if (context->COMMENT_CAP()) {
+      std::string c = context->COMMENT_CAP()->getText();
+      c.erase(0, 1);  // Remove the leading #
+      data << Str('#', c);
```

Comment:
I think it would be useful to have a header file listed the tags.  Ideally there would be a common set of tags for both the Python and C++ code, but simply having a mirror'd C++ header file and Python .py constant file would be very helpful.  As a concrete example, I believe this should be `Str(COMMENT_TAG, c)`, and somewhere above you have `constexpr char COMMENT_TAG = '#'`.

Reply:
I've added a rule in the Makefile to extract `TAG('c', long_name)` and produce a `tags.py` file with `long_name = b'c'`, and import this into `parse_fasm.py`.